pandas/tests/indexes/test_multi.py

```diff
@@ -3307,3 +3308,28 @@ def test_duplicate_multiindex_labels(self):
         with pytest.raises(ValueError):
             ind.set_levels([['A', 'B', 'A', 'A', 'B'], [2, 1, 3, -2, 5]],
                            inplace=True)
+
+    def test_use_enum_in_multiindex(self):
+        # GH 21298
+        # Allow use of Enums as one of the factors in a MultiIndex.
+        MyEnum = Enum("MyEnum", "A B")
+        df = pd.DataFrame(columns=pd.MultiIndex.from_product(iterables=[
+            MyEnum,
+            [1, 2]
+        ]))
+
+        # cf. https://github.com/pandas-dev/pandas/blob/
+        # 0c65c57a279e755ab7093db925d1e580f9878dae/pandas/util/testing.py#L793-L799
+        unique = df.columns.levels[0]
+        labels = df.columns.labels[0]
+        filled = take_1d(unique.values, labels, fill_value=unique._na_value)
+        df_index_0 = unique._shallow_copy(filled, name=df.columns.names[0])
+        exp_index_0 = pd.Index([MyEnum.A, MyEnum.A, MyEnum.B, MyEnum.B],
+                               dtype='object')
```

Comment:
this is pretty complicated, why cannot you just directly construct the Index here?